### PR TITLE
IMAGING-285 RationalNumber to support unsigned int format

### PR DESCRIPTION
**Pull request by kinow**
@gwlucastrig I created a new PR from your branch, but instead of using kinow/commons-imaging, picked apache/commons-imaging as base for the pull request.

Please take a look to see if I didn't do anything wrong. If not, I will start reviewing it :+1: 

Thanks!!
Bruno

### CODE REVIEW

File: src/main/java/org/apache/commons/imaging/formats/tiff/fieldtypes/FieldTypeRational.java [hunk 1]
```diff
@@ -31,11 +31,19 @@ public FieldTypeRational(final int type, final String name) {
     @Override
     public Object getValue(final TiffField entry) {
         final byte[] bytes = entry.getByteArrayValue();
+        boolean unsignedType = true;
+        if(entry.getFieldType()==RATIONAL){
+            unsignedType = true;
+        }else if(entry.getFieldType()==SRATIONAL){
+            unsignedType = false;
+        }
```
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/formats/tiff/fieldtypes/FieldTypeRational.java:39]: What about:

```suggestion
        boolean unsignedType = entry.getFieldType() != SRATIONAL;
```

File: src/main/java/org/apache/commons/imaging/common/RationalNumber.java [hunk 2]
```diff
@@ -20,6 +20,14 @@
...
 /**
  * Rational number, as used by the TIFF image format.
+ * <p>
+ * The TIFF format specifies two data types for rational numbers based on
+ * a pair of 32-bit integers.  Rational is based on unsigned 32-bit integers
+ * and SRational is based on signed 32-bit integers.  This treatment is
+ * problematic in Java because Java does not support unsigned types.
+ * To address this challenge, this class stores the numerator and divisor
+ * in long (64-bit) integers, applying masks as necessary for the unsigned
+ * type.
```
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/common/RationalNumber.java:30]: :+1: thanks for the useful comment!

File: src/main/java/org/apache/commons/imaging/common/RationalNumber.java [hunk 3]
```diff
@@ -73,8 +124,41 @@ private static long gcd(final long a, final long b) {
...
+     * @return a valid instance with a negated value.
+     */
     public RationalNumber negate() {
-        return new RationalNumber(-numerator, divisor);
+        long n = numerator;
+        long d = divisor;
+        if (unsignedType) {
+            // An instance of an unsigned type can be negated if and only if
+            // its high-order bit (the sign bit) is clear. If the bit is set,
+            // it is necessary to adjust the
```
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/common/RationalNumber.java:144]: to adjust the?

File: src/main/java/org/apache/commons/imaging/common/RationalNumber.java [hunk 4]
```diff
@@ -73,8 +124,41 @@ private static long gcd(final long a, final long b) {
         return gcd(b, a % b);
     }
 
+    /**
+     * Negates the value of the RationalNumber. If the numerator of this
+     * instance has its high-order bit set, then its value is too large
+     * to be treated as a Java 32-bit signed integer.  In such a case, the
+     * only way that a RationalNumber instance can be negated is to
+     * sacrifice one bit of precision by dividing both the numerator and
+     * the denominator by 2. However, if that computation would result
```
Comment (kinow) [inline src/main/java/org/apache/commons/imaging/common/RationalNumber.java:133]: by `2` or by the greatest common divisor?

**Comment by kinow** [comment]
ping @gwlucastrig

File: src/main/java/org/apache/commons/imaging/formats/tiff/fieldtypes/FieldTypeRational.java (hunk 1 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/fieldtypes/FieldTypeRational.java:39]: That will work.  There are only the two conditions.

File: src/main/java/org/apache/commons/imaging/common/RationalNumber.java (hunk 3 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/common/RationalNumber.java:144]: numerator and denominator values.

File: src/main/java/org/apache/commons/imaging/common/RationalNumber.java (hunk 4 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/common/RationalNumber.java:133]: by the computed greatest common denominator.

**Comment by gwlucastrig** [comment]
Is the plan for you to make these changes or should I.

**Comment by kinow** [comment]
I think it'd be better if you did these changes @gwlucastrig . I only created the PR to change the base (it was kinow/commons-imaging).

**Comment by kinow** [comment]
While on it, you can rebase/squash too :grimacing: then it should be ready to merge.

**Comment by gwlucastrig** [comment]
I will make the recommended (and completely sensible) changes, rebase, and submit within the next couple of days.

File: src/main/java/org/apache/commons/imaging/common/RationalNumber.java (hunk 2 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/common/RationalNumber.java:30]: You're welcome as always...

File: src/main/java/org/apache/commons/imaging/common/RationalNumber.java (hunk 4 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/common/RationalNumber.java:133]: done

File: src/main/java/org/apache/commons/imaging/common/RationalNumber.java (hunk 3 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/common/RationalNumber.java:144]: done

File: src/main/java/org/apache/commons/imaging/formats/tiff/fieldtypes/FieldTypeRational.java (hunk 1 above)
Comment (gwlucastrig) [reply src/main/java/org/apache/commons/imaging/formats/tiff/fieldtypes/FieldTypeRational.java:39]: done

**Comment by gwlucastrig** [comment]
Well, predictably, somethings gone wrong with my branch and pull request...   I made the changes, did the squash, and then a push...  But they did not seem to make it into the pull request.   In my git shell, I got the following:

remote: Resolving deltas: 100% (1218/1218), completed with 185 local objects.
remote:
remote: Create a pull request for 'IMAGING-285' on GitHub by visiting:
remote:      https://github.com/gwlucastrig/commons-imaging/pull/new/IMAGING-285
remote:
To https://github.com/gwlucastrig/commons-imaging.git
 * [new branch]        IMAGING-285 -> IMAGING-285

**Comment by kinow** [comment]
@gwlucastrig I think we both learned that GitHub must keep some sort of internal unique ID per fork. Your previous fork has disappeared, hence the new PR. Let's create a new PR and close this one :+1: I will consult my previous comments while reviewing the new PR.

**Pull request closed without merging**